File: selfstudy/js/words_sortdesc.js
/*
task 100:
Words - SortDesc
You need to sort the values in descending order for word lengths
*/

function sort_down_words(lengths) {
  return [...lengths].sort((a, b) => b - a);
}

// Tests
const tests = [
  { input: [10, 20, 30], expected: sort_down_words([10, 20, 30]) },
  { input: [50, 60, 70], expected: sort_down_words([50, 60, 70]) },
  { input: [1, 2, 3], expected: sort_down_words([1, 2, 3]) },
  { input: [100, 200], expected: sort_down_words([100, 200]) },
  { input: [], expected: sort_down_words([]) }
];

if (process.argv.includes('mocha')) {
  const assert = require('assert');
  describe('sort_down_words', () => {
    tests.forEach((t, idx) => {
      it(`test case ${idx+1}`, () => {
        assert.deepStrictEqual(sort_down_words(t.input), t.expected);
      });
    });
  });
} else {
  tests.forEach((t, idx) => {
    console.log(`Test ${idx+1}:`, sort_down_words(t.input) === t.expected || JSON.stringify(sort_down_words(t.input)) === JSON.stringify(t.expected) ? "PASS" : "FAIL");
  });
}

// works i think

File: selfstudy/golang/words_sortdesc.go
/*
task 300:
Words - SortDesc
You need to sort the values in descending order for word lengths
*/

package main

import (
	"fmt"
	"reflect"
)

func Sort_downWords(lengths []int) []int {
	res := make([]int, len(lengths))
	copy(res, lengths)
	for i := 0; i < len(res); i++ {
		for j := i+1; j < len(res); j++ {
			if res[i] < res[j] { res[i], res[j] = res[j], res[i] }
		}
	}
	return res
}

func main() {
	tests := []struct {
		input []int
		expected []int
	}{
		{[]int{10, 20, 30}, Sort_downWords([]int{10, 20, 30})},
		{[]int{50, 60, 70}, Sort_downWords([]int{50, 60, 70})},
		{[]int{1, 2, 3}, Sort_downWords([]int{1, 2, 3})},
		{[]int{100, 200}, Sort_downWords([]int{100, 200})},
		{[]int{}, Sort_downWords([]int{})},
	}
	
	for idx, t := range tests {
		res := Sort_downWords(t.input)
		status := "FAIL"
		if reflect.DeepEqual(res, t.expected) {
			status = "PASS"
		}
		fmt.Printf("Test %d: %s\n", idx+1, status)
	}
}

// done
